### ISSUE TITLE
T5 model: There were missing keys in the checkpoint model loaded: ['encoder.embed_tokens.weight', 'decoder.embed_tokens.weight', 'lm_head.weight'].

### ISSUE THREAD

**Issue by alexcoca** [huggingface/transformers#27972]
### System Info

- `transformers` version: 4.35.2
- Platform: Linux-5.4.0-148-generic-x86_64-with-glibc2.27
- Python version: 3.10.11
- Huggingface_hub version: 0.19.4
- Safetensors version: 0.4.0
- Accelerate version: 0.24.1
- Accelerate config: 	not found
- PyTorch version (GPU?): 1.13.1+cu117 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: yes (RTX3090)
- Using distributed or parallel set-up in script? no

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

Steps to reproduce.

1. Run any transformer example fine-tuning a t5 model (I am using `Salesforce/codet5p-220m` but the issue can probably be reproduced with other T5 models (certainly FlanT5)
2. Stop the trainer 
3. Restart the training using the `restart_from_chekpoint=True` CLI option and setting `output_dir` to be the checkpoint directory (ie  where the `checkpoint-[step]` directories are created)
4. Observe the warning:

[WARNING|trainer.py:2231] 2023-12-12 11:09:58,921 >> There were missing keys in the checkpoint model loaded: ['encoder.embed_tokens.weight', 'decoder.embed_tokens.weight', 'lm_head.weight'].

### Expected behavior

Either there is no warning or the warning message tells the user if the warning applies to them. My intuition here is that nothing is wrong: I am using `T5ForConditionlGeneration` out of the box (so no `lm_head`) and the encoder and decoder enmbedings are tied (and hopefully loaded ?!). Is this a case of extending the warning to be more explicit?

@younesbelkada

**Comment by wuyuhan111111** [comment]
I also have this issue, and once prompted, the training will be terminated directly. Have you resolved it?

**Comment by amyeroberts** [comment]
cc @muellerzr @pacman100 as the warning seems to be coming from trainer

**Comment by valentas-kurauskas** [comment]
I also get with [run_summarizaton.py](https://github.com/huggingface/transformers/blob/main/examples/pytorch/summarization/run_summarization.py) and `--model_name_or_path "google/mt5-base"`
```
.. missing keys in the checkpoint model loaded: ['encoder.embed_tokens.weight', 'decoder.embed_tokens.weight']
```
But fine-tuning continues from the last checkpoint rather than crashing. However, `eval_loss` increases for the next checkpoint after restart, suggesting these weights are important and are really not saved/reloaded.

**Comment by muellerzr** [comment]
Related to https://github.com/huggingface/transformers/issues/27293

**Comment by valentas-kurauskas** [comment]
@muellerzr thanks for linking to the issue. But the solution mentioned there is for `accelerate`, and in this case I have a problem with checkpoints saved by Trainer.

**Comment by palit-dev** [comment]
Facing the same issue for all T5 as well as RoBERTa models. Any solution yet?

**Comment by alexcoca** [comment]
@muellerzr and @pacman100 - it's slightly concerning that this warning still appears and we don't have any dialogue about whether this is a true issue or a fluke. Is there any understanding of what transformers release guarantees correct checkpoint loading? I have used the library to implement my next research paper, but I don't know whether or not I can actually use any of the models given the warning on model loading? Let's chat and see how we can get to the bottom of this.

**Comment by muellerzr** [comment]
@alexcoca can you give us a full clean reproducer please? That's the best way we can help. (Cc @Narsil)

**Comment by AwaysAbdiwahid** [comment]
also, I had a similar issue with training the Bart model for abstractive-based text summarization.
but I thought that the problem was caused by how I initialized my model because I configured the model using BartConfig class and also ByteLevelBPETokenizer for text tokenization. but I am wonder if this issue has an impact of the model performance

**Comment by preeti-chauhan** [comment]
Using mt5: There were missing keys in the checkpoint model loaded: ['encoder.embed_tokens.weight', 'decoder.embed_tokens.weight']. Is this unresolved/expected? (although, the next epoch loss is less than previous.)

**Comment by TopCoder2K** [comment]
I'm facing a similar warning with `DistilBertForMaskedLM`:
```
There were missing keys in the checkpoint model loaded: ['vocab_projector.weight'].
```
I'm saving with
```
trainer.save_model(dir)
```
and loading with
```
AutoModelForMaskedLM.from_pretrained(dir).to(self.device)
```
I have encountered several times that setting `save_safetensors=False` helps. Is this the right solution if I don't need the safetensors format, @muellerzr?

**Comment by thistlillo** [comment]
+1 here on January 31, 2025.

**Comment by sebastian-montero** [comment]
Closed? Im facing this issue as well..

**Comment by JamePeng** [comment]
+2 here on Feb. 24, 2025.

**Comment by seungeonlee-enerzai** [comment]
+3 here on Feb. 27, 2025.

**Comment by Rocketknight1** [comment]
I'll reopen, but can someone please post a simple complete reproducer to make things easier on the team? cc @muellerzr @SunMarc

**Comment by JamePeng** [comment]
> I'll reopen, but can someone please post a simple complete reproducer to make things easier on the team? cc [@muellerzr](https://github.com/muellerzr) [@SunMarc](https://github.com/SunMarc)

Just use the trainer of transformer for finetune the llama3.2-3b, it will be happen

**Comment by ruiruiq** [comment]
same problem in switchtransformer

**Comment by ruiruiq** [comment]
@abdulfatir

> [@muellerzr](https://github.com/muellerzr) it looks like some of these warnings are coming from here:
> 
> [transformers/src/transformers/trainer.py](https://github.com/huggingface/transformers/blob/65b8e38aac2a2e798d5d7406a970a8f206815b1c/src/transformers/trainer.py#L3039-L3050)
> 
> Lines 3039 to 3050 in [65b8e38](/huggingface/transformers/commit/65b8e38aac2a2e798d5d7406a970a8f206815b1c)
> 
>  def _issue_warnings_after_load(self, load_result): 
>      if len(load_result.missing_keys) != 0: 
>          if self.model._keys_to_ignore_on_save is not None and set(load_result.missing_keys) == set( 
>              self.model._keys_to_ignore_on_save 
>          ): 
>              self.model.tie_weights() 
>          else: 
>              logger.warning(f"There were missing keys in the checkpoint model loaded: {load_result.missing_keys}.") 
>      if len(load_result.unexpected_keys) != 0: 
>          logger.warning( 
>              f"There were unexpected keys in the checkpoint model loaded: {load_result.unexpected_keys}." 
>          ) 
> This compares `load_result.missing_keys` with `self.model._keys_to_ignore_on_save` and I am not completely sure that that's the right thing to do. Taking the example of T5, the right comparison would be with `_tied_weights_keys`:
> 
> [transformers/src/transformers/models/t5/modeling_t5.py](https://github.com/huggingface/transformers/blob/65b8e38aac2a2e798d5d7406a970a8f206815b1c/src/transformers/models/t5/modeling_t5.py#L1710-L1714)
> 
> Lines 1710 to 1714 in [65b8e38](/huggingface/transformers/commit/65b8e38aac2a2e798d5d7406a970a8f206815b1c)
> 
>  class T5ForConditionalGeneration(T5PreTrainedModel, GenerationMixin): 
>      _keys_to_ignore_on_load_unexpected = [ 
>          "decoder.block.0.layer.1.EncDecAttention.relative_attention_bias.weight", 
>      ] 
>      _tied_weights_keys = ["encoder.embed_tokens.weight", "decoder.embed_tokens.weight", "lm_head.weight"]

[INFO|trainer.py:2733] 2025-03-13 20:17:14,684 >> Loading model from ./switch_transformers/summarization/retrain/checkpoint-7000.
[WARNING|trainer.py:2976] 2025-03-13 20:17:15,486 >> There were missing keys in the checkpoint model loaded: ['encoder.embed_tokens.weight', 'decoder.embed_tokens.weight', 'lm_head.weight'].

so how i can jump the warning? 👍

**Comment by abdulfatir** [comment]
I think `transformers` needs to fix how missing keys are checked. If you're using a fully custom model, you could set the `_keys_to_ignore_on_save` attribute in your model to the know missing keys.

**Comment by long8v** [comment]
same here. it seems to happen in this line https://github.com/huggingface/transformers/blob/51f94ea06d19a6308c61bbb4dc97c40aabd12bad/src/transformers/modeling_utils.py#L4832-L4835 

when lm_head happens to save other than 0th safetensor, it cannot be loaded properly. 

<img width="824" height="437" alt="Image" src="https://github.com/user-attachments/assets/b4295bfa-7a8f-4012-b149-cedf20f429e4" />
